mypy/plugin.py

```diff
@@ -379,3 +421,346 @@ def int_pow_callback(ctx: MethodContext) -> Type:
+                        # This decorator lets you set a default after the fact.
+                        own_attrs[func_decorator.expr.name].has_default = True
+
+                    if func_decorator.name in ('default', 'validator'):
+                        # These are decorators on the attrib object that only exist during
+                        # class creation time.  In order to not trigger a type error later we
+                        # just remove them.  This might leave us with a Decorator with no
+                        # decorators (Emperor's new clothes?)
+                        # TODO: Any way to type-check these?
+                        #       default should be Callable[[], T]
+                        #       validator should be Callable[[Any, 'Attribute', T], Any]
+                        #       where T is the type of the attribute.
+                        remove_me.append(func_decorator)
+
+            for dec in remove_me:
+                stmt.decorators.remove(dec)
+
+    taken_attr_names = set(own_attrs)
+    super_attrs = []
+
+    # Traverse the MRO and collect attributes.
+    for super_info in info.mro[1:-1]:
+        if super_info in attr_classes:
+            for a in attr_classes[super_info]:
+                # Only add an attribute if it hasn't been defined before.  This
+                # allows for overwriting attribute definitions by subclassing.
+                if a.name not in taken_attr_names:
+                    super_attrs.append(a)
+                    taken_attr_names.add(a.name)
+
+    attributes = super_attrs + list(own_attrs.values())
+    # Save the attributes so that subclasses can reuse them.
+    # TODO: Fix caching.
+    attr_classes[info] = attributes
+
+    if ctx.api.options.disallow_untyped_defs:
+        for attribute in attributes:
+            if attribute.type is None:
+                # This is a compromise.  If you don't have a type here then the __init__ will
+                # be untyped. But since the __init__ method doesn't have a line number it's
+                # difficult to point to the correct line number.  So instead we just show the
+                # error in the assignment, which is where you would fix the issue.
+                ctx.api.fail(messages.NEED_ANNOTATION_FOR_VAR, attribute.context)
+
+    # Check the init args for correct default-ness.  Note: This has to be done after all the
+    # attributes for all classes have been read, because subclasses can override parents.
+    last_default = False
+    for attribute in attributes:
+        if not attribute.has_default and last_default:
+            ctx.api.fail(
+                "Non-default attributes not allowed after default attributes.",
+                attribute.context)
+        last_default = attribute.has_default
+
+    self_type = fill_typevars(info)
+    function_type = ctx.api.named_type('__builtins__.function')
+
+    def add_method(method_name: str, args: List[Argument], ret_type: Type,
```

Comment:
Can you move this function to the outer scope? It looks like it is useful on its own (i.e., can be used by other plugins in future). Also please add a TODO to re-use the same code for `NamedTuple`.